pkg/resolution/resolver/git/resolver.go

```diff
@@ -156,29 +156,67 @@ func (r *Resolver) resolveAPIGit(ctx context.Context, params map[string]string)
 		return nil, fmt.Errorf("failed to create SCM client: %w", err)
 	}
 
-	content, _, err := scmClient.Contents.Find(ctx, fmt.Sprintf("%s/%s", params[orgParam], params[repoParam]), params[pathParam], params[revisionParam])
+	orgRepo := fmt.Sprintf("%s/%s", params[orgParam], params[repoParam])
+	path := params[pathParam]
+	ref := params[revisionParam]
+
+	// fetch the actual content from a file in the repo
+	content, _, err := scmClient.Contents.Find(ctx, orgRepo, path, ref)
 	if err != nil {
 		return nil, fmt.Errorf("couldn't fetch resource content: %w", err)
 	}
 	if content == nil || len(content.Data) == 0 {
-		return nil, fmt.Errorf("no content for resource in %s/%s %s", params[orgParam], params[repoParam], params[pathParam])
+		return nil, fmt.Errorf("no content for resource in %s %s", orgRepo, path)
 	}
 
-	repo, _, err := scmClient.Repositories.Find(ctx, fmt.Sprintf("%s/%s", params[orgParam], params[repoParam]))
+	commitSHA := content.Sha // content sha is different from commit sha
+	// TODO: fetch actual commit sha based on the provided revision which can be commitsha/branch/tag.
+	// commitSHA, err := fetchCommitSHAForAPIGit(ctx, scmClient, orgRepo, ref)
+	// if err != nil {
+	// 	return nil, fmt.Errorf("cannot find the actual commit sha: %w", err)
+	// }
+
+	// fetch the repository URL
+	repo, _, err := scmClient.Repositories.Find(ctx, orgRepo)
 	if err != nil {
 		return nil, fmt.Errorf("couldn't fetch repository: %w", err)
 	}
 
 	return &resolvedGitResource{
 		Content:  content.Data,
-		Revision: content.Sha,
+		Revision: commitSHA,
 		Org:      params[orgParam],
 		Repo:     params[repoParam],
 		Path:     content.Path,
 		URL:      repo.Clone,
 	}, nil
 }
 
+// fetchCommitSHAForAPIGit fetches actual commit SHA using scm api based on
+// the user provided revision which can be the name of commitSHA/branch/tag.
+func fetchCommitSHAForAPIGit(ctx context.Context, scmClient *scm.Client, orgRepo, revision string) (string, error) {
+	// check if the provided revision is an actual commit.
+	c, _, err := scmClient.Git.FindCommit(ctx, orgRepo, revision)
+
+	if err == nil && c != nil {
+		return c.Sha, nil
+	}
+
+	// find the commit sha if the provided revision is a tag name
+	release, _, err := scmClient.Releases.FindByTag(ctx, orgRepo, revision)
+	if err == nil && release != nil {
+		return release.Commitish, nil
+	}
+
+	// find the commit sha if the provided revision is a branch name
+	ref, _, err := scmClient.Git.FindBranch(ctx, orgRepo, revision)
+	if err == nil && ref != nil {
+		return ref.Sha, nil
+	}
+
+	return "", fmt.Errorf("the provided revision %q is not a real commit sha or a name of a branch or a tag.", revision)
+}
+
```

Comment:
Hi @abayer ,
This is for scm cloning approach only. What I am trying to do here is retrieve the actual commit sha rather than content sha regardless of the user-provided revision param which can be commit sha / branch name / tag name.

The approach in `fetchCommitSHAForAPIGit` seems the best solution for now but the fake test utilities in the scm package do no implement GitService's [FindBranch](https://github.com/jenkins-x/go-scm/blob/648bf6dd5dd722041d215869a547045f20984d29/scm/driver/fake/git.go#L37-L39), which blocks unit test in our case. 

The alternative I tried is to use GitService's [FindRef](https://github.com/jenkins-x/go-scm/blob/338e0fcd608a59f137b4aa9f97d4761082cf8729/scm/git.go#L93-L94) to find commit sha, but this is not optional because the 3rd argument expected differs from provider to provider. For example, [github](https://docs.github.com/en/rest/git/refs#get-a-reference)
- we need to prepend "heads/" to the revision if it's branch name
- we need to prepend "tags/"  to the revision if it's tag name 


wdyt of the approach in `fetchCommitSHAForAPIGit`? If it sounds good, I am happy to submit a PR to implement the missing fake [FindBranch](https://github.com/jenkins-x/go-scm/blob/648bf6dd5dd722041d215869a547045f20984d29/scm/driver/fake/git.go#L37-L39) in scm package to help us do unit test.

Reply:
I'd say go with adding the fake `FindBranch` - cc me on the PR and I'll review it and ping the maintainer to make sure it gets merged. =)

Reply:
Thanks for the thoughts @abayer.

@wlynch just helped me realize the git service's [FindCommit](https://pkg.go.dev/github.com/jenkins-x/go-scm/scm#GitService) in scm api can also fetch a commit by commit sha/branch/tag name, which I've used in `fetchCommitSHAForAPIGit` but I thought it only recognizes an commit sha but not tag/branch name :D